src/auth/firebase.service.ts

```diff
@@ -0,0 +1,156 @@
+          databaseURL: options.databaseUri,
+        }
+      : {};
+
+    this.app = initializeApp(appOptions);
+    this.db = getFirestore(this.app);
+  }
+
+  async verifyIdToken(idToken: string) {
+    return getAuth(this.app).verifyIdToken(idToken);
+  }
+
+  async getUserInfo(uid: string) {
+    return this.db.collection('users').doc(uid).get();
+  }
+
+  private upsertElementByArticleId = (
+    articleId: string,
+    list: any[],
+    element: any,
+  ) => {
+    const newArray = [...list];
+    const index = list.findIndex((s) => s.articleId === articleId);
+    index !== -1 ? (newArray[index] = element) : newArray.push(element);
+    return newArray;
+  };
+
+  async updateArticleRecord(uid: string, articleId: string) {
+    const userRef = this.db.collection('articles').doc(uid);
+    const doc = await userRef.get();
+
+    const editedList = doc.get('edited') ?? [];
+    const subscribedList = doc.get('subscribed') ?? [];
+    const element = {
+      articleId,
+      timeStamp: Timestamp.now(),
+    };
+
+    const newEditedList = this.upsertElementByArticleId(
+      articleId,
+      editedList,
+      element,
+    );
+    const newSubscribedList = this.upsertElementByArticleId(
+      articleId,
+      subscribedList,
+      element,
+    );
+
+    userRef.set(
+      { edited: newEditedList, subscribed: newSubscribedList },
+      { merge: true },
+    );
+
+    return Promise.resolve({
+      edited: newEditedList,
+      subscribed: newSubscribedList,
+    });
+  }
+
+  async updateArticleSubscription(
+    uid: string,
+    articleId: string,
+    isSubscribe: boolean,
+  ) {
+    const userRef = this.db.collection('articles').doc(uid);
+    const doc = await userRef.get();
+
+    const subscribedList = doc.get('subscribed') ?? [];
+
+    let newSubscribedList = [];
+
+    if (isSubscribe) {
+      const element = {
+        articleId,
+        timeStamp: Timestamp.now(),
+      };
+      newSubscribedList = this.upsertElementByArticleId(
+        articleId,
+        subscribedList,
+        element,
+      );
+    } else {
+      newSubscribedList = subscribedList.filter(
+        (s: any) => s.articleId !== articleId,
+      );
+    }
+
+    userRef.set({ subscribed: newSubscribedList }, { merge: true });
+
+    return Promise.resolve(newSubscribedList);
+  }
+
+  async accumulateUserPoints(uid: string, point: number) {
+    const userRef = this.db.collection('users').doc(uid);
+    const doc = await userRef.get();
+
+    const origPoint: number = doc.get('points') ?? 0;
+    const newPoint = origPoint + point;
+    userRef.set({ point: newPoint }, { merge: true });
+
+    return Promise.resolve(newPoint);
+  }
+
+  async getUserPoints(uid: string) {
+    const userRef = this.db.collection('users').doc(uid);
+    const doc = await userRef.get();
+
+    const point: number = doc.get('points') ?? 0;
+
+    return Promise.resolve(point);
+  }
+
+  async updateNotification(uid: string, notifications: any[], maxRemain = 20) {
```

Comment:
maxRemain should write in const value as DEFAULT_MAX_REMAIN on the top of this file.

And does notification has its type? Not good in use any as type

Reply:
done.